### PR TITLE
Nautobot Docs Revamp

### PR DESCRIPTION
**Pull request by cmsirbu**
# What's Changed

- Changed indent to 2 spaces in `mkdocs.yml` and made it yamllint compliant
- Added `assets` folder for theme related assets and overrides
    + Leaving nautobot logos in `docs` for now to avoid any breakage if they are linked to
- Changed theme to "Material for mkdocs" with NTC customization and branding
- Revamped landing page (root `index.md`) to include an overview of the documentation structure.
- Added the `Nautobot Apps` section to showcase and link to other documentation sites.
- Added new docs dependencies to poetry dev-dependencies
- Tested that docs still build in the development environment - hosted docs look good, model docs look good.
- Fixed markdownlint errors, both old and new.
- Updated example_plugin docs to match the same theme and tested embedded hosting.

A live demo of how this looks like can be found [here](https://nbatest-ml.readthedocs.io/projects/core/en/latest/).

The goal of this first PR is to port the docs without any structural changes or major updates to content, mostly to prevent any possible link breakage. The following step after this PR is merged will be to swap the project domain over to `docs.nautobot.com/projects/core/`. Once the dust has settled I think we can do larger content changes.

Screenshot from 2022-08-31 11-40-24

# TODO

- [x] Explanation of Change(s)
- [x] Linked live build for review
- [x] Example Plugin Updates (when adding/changing features)
- [ ] App logos to be updated before merge with consistent sizing and style

### CODE REVIEW

File: mkdocs.yml
Comment (glennmatthews) [inline mkdocs.yml:6]: To address #2345, this *maybe* should be `https://nautobot.readthedocs.io/en/latest/` or `https://nautobot.readthedocs.io/en/stable/`. I see that the 404 page rendering is broken on https://nbatest-ml.readthedocs.io/projects/core/en/latest/development/404, although it seems to be working on https://nautobot--2347.org.readthedocs.build/en/2347/development/404...

File: mkdocs.yml
Comment (glennmatthews) [inline mkdocs.yml:48]: Looks like this PR was merged 27 days ago.

File: mkdocs.yml
Comment (glennmatthews) [inline mkdocs.yml:168]: These two probably should be moved into the `Development Guide` top-level section

File: mkdocs.yml
Comment (glennmatthews) [inline mkdocs.yml:180]: Perhaps "Developer Guide"?

File: mkdocs.yml
Comment (glennmatthews) [inline mkdocs.yml:97]: I recognize that this is mostly just porting over the existing documentation structure as-is, but while we're touching this, I'd be tempted to make the following changes:

- Move `docker/index.md` into the `Installation` heading rather than having a separate top-level heading with only one item under it.
- Move `Administration` up to between `External Authentication` and `User Guides`
- Combine `User Guides` and `Getting Started in the Web UI` into a single `User Guides` section, starting with the `Getting Started` pages and ending with the `User Guides` pages
- Make `Plugins` the last item under `Documentation` instead of having it come before `REST API` and `GraphQL`.

File: nautobot/docs/apps/index.md
Comment (glennmatthews) [inline nautobot/docs/apps/index.md:5]: Perhaps "Community-Developed Nautobot Apps"?

File: nautobot/docs/apps/index.md
Comment (glennmatthews) [inline nautobot/docs/apps/index.md:9]: Perhaps "Network to Code Nautobot Apps"?

File: nautobot/docs/apps/index.md
Comment (glennmatthews) [inline nautobot/docs/apps/index.md:20]: Would be good to have a consistent style (verb tense, phrasing, etc.) for the Description of each app. "Automate ..." and "Simplify..." versus "This app does..." versus "Collection of models..." versus "Adds a chatbot..."

File: pyproject.toml
Comment (glennmatthews) [inline pyproject.toml:169]: Looks like these two dependencies should be bumped to ==3.6.1 and ==1.3.1 respectively.

File: pyproject.toml
Comment (glennmatthews) [inline pyproject.toml:171]: Should this be ==8.4.2 instead of ~8.4.2? Since the docs/requirements.txt is expecting an exact version.

**Review by glennmatthews — COMMENTED** [review]
Awesome stuff!

File: mkdocs.yml
Comment (cmsirbu) [reply mkdocs.yml:48]: Sadly last I checked they decided to not inject jquery still. There's been some more work done it seems at removing dependency on jquery, I'll try a build without it and see if the flyout still works.

https://github.com/readthedocs/readthedocs.org/pull/9359#issuecomment-1205650777

File: nautobot/docs/apps/index.md
Comment (cmsirbu) [reply nautobot/docs/apps/index.md:20]: Agree, it's a hodge-podge taken from the READMEs and documentation of each app currently. Ideally each app would have a single-paragraph description (like an abstract) that we'd replicate here.

File: pyproject.toml
Comment (cmsirbu) [reply pyproject.toml:169]: The `~` takes care of that, but since `poetry.lock` will influence what gets installed, I'm not sure where to stand on the exact versions vs this slightly flexible "any-later-patch" is OK.

File: mkdocs.yml
Comment (cmsirbu) [reply mkdocs.yml:97]: Yeah the goal of this first PR is to port the docs without any structural changes or major updates to content, mostly to prevent any possible breakage. The following step after the PR is merged will be to swap domain over to `docs.nautobot.com` so I'd like to reduce the amount of potential broken links at this stage. I'll update this in the PR summary, thanks for the reminder.

Changes to the nav (like the suggested move of the developing plugins bit earlier) should be fine.

File: .yamllint.yml
Comment (bryanculver) [inline .yamllint.yml:12]: Is this for your local development? I see no tooling/CI changes that will look at this.

Comment (cmsirbu) [reply .yamllint.yml:12]: I used it to validate the YAML I touched was good. Not adding to CI since it has a problem with a lot of other files and I don't want to scope-creep this PR. I'll remove it and perhaps do a separate PR if adding yamllint is wanted?

Comment (bryanculver) [reply .yamllint.yml:12]: > I used it to validate the YAML I touched was good. Not adding to CI since it has a problem with a lot of other files and I don't want to scope-creep this PR. I'll remove it and perhaps do a separate PR if adding yamllint is wanted?

This would be preferred, as at present it would require optional setup in development environments with no enforcement in CI.

File: mkdocs.yml
Comment (cmsirbu) [reply mkdocs.yml:6]: Setting the `site_url` to the full `https://nautobot.readthedocs.io/en/stable/` does the trick. The CI RTD build for the PR is a fluke and the fact that it is working without the fix actually worries me that it's not representative of the production setup :(

For future reference - there is an [open issue on RTD](https://github.com/readthedocs/readthedocs.org/issues/4820) about injecting the correct canonical tag in mkdocs builds, but it seems to be stalled.

File: mkdocs.yml
Comment (cmsirbu) [reply mkdocs.yml:48]: I tested and we still need to inject jquery... it seems RTD are working on removing their dependency on jquery, but in the meantime decided to not inject jquery together with their code.

File: mkdocs.yml
Comment (cmsirbu) [reply mkdocs.yml:97]: Nav updated with suggestions.

**Comment by cmsirbu** [comment]
While this PR addresses #2345, it won't apply to any tagged versions that are older than when this gets released. Any **new** deployments won't start working until `en/stable` is updated (which is considered the canonical version), even if `en/latest` or other tagged builds are already live with the new theme.

**Review by glennmatthews — APPROVED** [review]
Terrific stuff!

**Comment by cmsirbu** [comment]
In golden-config, @itdependsnetworks added a basic test to compare the `docs/requirements.txt` versions against the ones added to `pyproject.toml`. It works with exact pinned versions and is great for preventing any drift due to the duplication of the docs dependencies. Is this something we can fit in here as well? If so, what's the best place to put it?

Test: https://github.com/nautobot/nautobot-plugin-golden-config/blob/develop/nautobot_golden_config/tests/test_basic.py

CC @glennmatthews @bryanculver @lampwins

File: pyproject.toml
Comment (cmsirbu) [reply pyproject.toml:174]: Not used yet, but it's expected functionality IIRC, so I added them for consistency of docs capabilities over the ecosystem.

**Comment by glennmatthews** [comment]
> In golden-config, @itdependsnetworks added a basic test to compare the `docs/requirements.txt` versions against the ones added to `pyproject.toml`. It works with exact pinned versions and is great for preventing any drift due to the duplication of the docs dependencies. Is this something we can fit in here as well? If so, what's the best place to put it?

I'm not opposed to this. Probably putting it somewhere like `nautobot/core/tests/test_docs_requirements.py` would be appropriate.

File: pyproject.toml
Comment (glennmatthews) [reply pyproject.toml:174]: Fair enough. I know we've talked about adding autogenerated source documentation to the core project, but I don't know that we have a tracking issue open for that yet.

**Comment by bryanculver** [comment]
> > In golden-config, @itdependsnetworks added a basic test to compare the `docs/requirements.txt` versions against the ones added to `pyproject.toml`. It works with exact pinned versions and is great for preventing any drift due to the duplication of the docs dependencies. Is this something we can fit in here as well? If so, what's the best place to put it?
> 
> I'm not opposed to this. Probably putting it somewhere like `nautobot/core/tests/test_docs_requirements.py` would be appropriate.

@cmsirbu I won't hold up this PR on this because we already get some consistency for these files via Renovate. Feel free to open a housekeeping issue to add said test.

**Review by bryanculver — APPROVED** [review]

**Pull request merged**
